horovod/spark/common/util.py

```diff
@@ -0,0 +1,518 @@
+
+    metadata = dict()
+    for field in df.schema.fields:
+        col = field.name
+        col_types = all_col_types[col].copy()
+
+        if DenseVector in col_types and SparseVector in col_types:
+            # If a col has DenseVector type (whether it is mixed sparse and dense vector or just
+            # DenseVector), convert all of the values to dense vector
+            is_sparse_vector_only = False
+            spark_data_type = DenseVector
+            convert_to_target = constants.ARRAY
+        elif DenseVector in col_types:
+            # If a col has DenseVector type (whether it is mixed sparse and dense vector or just
+            # DenseVector), convert all of the values to dense vector
+            is_sparse_vector_only = False
+            spark_data_type = DenseVector
+            convert_to_target = constants.ARRAY
+        elif SparseVector in col_types:
+            # If a col has only sparse vectors, convert all the data into custom dense vectors
+            is_sparse_vector_only = True
+            spark_data_type = SparseVector
+            convert_to_target = constants.CUSTOM_SPARSE
+        else:
+            is_sparse_vector_only = False
+            spark_data_type = type(field.dataType)
+            convert_to_target = constants.NOCHANGE
+
+        # Explanation of the fields in metadata
+        #     dtype:
+        #
+        #     spark_data_type:
+        #         The spark data type from dataframe schema: type(field.dataType). If column has
+        #         mixed SparseVector and DenseVector we categorize it as DenseVector.
+        #
+        #     is_sparse_vector_only:
+        #         If all the rows in the column were sparse vectors.
+        #
+        #     shape:
+        #         Determines the shape of the data in the spark dataframe. It is useful for sparse
+        #         vectors.
+        #
+        #     intermediate_format:
+        #         Specifies if the column need to be converted to a different format so that
+        #         petastorm can read it. It can be one of ARRAY, CUSTOM_SPARSE, or NOCHANGE. It is
+        #         required because petastorm cannot read DenseVector and SparseVectors. We need to
+        #         identify these types and convert them to petastorm compatible type of array.
+
+        metadata[col] = {'spark_data_type': spark_data_type,
+                         'is_sparse_vector_only': is_sparse_vector_only,
+                         'shape': col_shapes[col],
+                         'intermediate_format': convert_to_target,
+                         'max_size': col_max_sizes[col]}
+
+    return metadata
+
+
+def to_petastorm_fn(schema_cols, metadata):
+    ARRAY = constants.ARRAY
+    CUSTOM_SPARSE = constants.CUSTOM_SPARSE
+
+    # Convert Spark Vectors into arrays so Petastorm can read them
+    def to_petastorm(row):
+        import numpy as np
+        from pyspark import Row
+
+        fields = row.asDict().copy()
```

Comment:
When all columns are `NOCHANGE`, this is really hurting. Should be improved as the next step.

Reply:
Good point.  Added TODO for followup PR to remove this when schema is all scalar.